From d8432ca288eef1043587e3004fb3a7b25cff4020 Mon Sep 17 00:00:00 2001
From: Ruomio <1065940593@qq.com>
Date: Wed, 22 Jan 2025 15:41:04 +0800
Subject: [PATCH 06/23] soc: sdram

---
 perip/sdram/sdram.v | 77 +++++++++++++++++++++++----------------------
 1 file changed, 39 insertions(+), 38 deletions(-)

diff --git a/perip/sdram/sdram.v b/perip/sdram/sdram.v
index ce707858..434adffb 100644
--- a/perip/sdram/sdram.v
+++ b/perip/sdram/sdram.v
@@ -10,14 +10,14 @@ module sdram(
   input [ 1:0] dqm,
   inout [15:0] dq
 );
-  import "DPI-C" function void sdram_read(input byte ba, input int row_addr, input int col_addr, input shortint wstrb, output shortint rdata);
-  import "DPI-C" function void sdram_write(input byte ba, input int row_addr, input int col_addr, input shortint wstrb, input shortint wdata);
+  import "DPI-C" function void sdram_read(input byte ba, input int row_addr, input int col_addr, input shortint wstrb, output int rdata);
+  import "DPI-C" function void sdram_write(input byte ba, input int row_addr, input int col_addr, input shortint wstrb, input int wdata);
 
   // Internal registers and wires
   // reg [15:0] memory [0:262143]; // 256Mb SDRAM (16M x 16)
   reg [15:0] dq_out;
   reg dq_en;
-  reg [12:0] row_addr;
+  reg [12:0] row_addr [3:0];
   reg [1:0] bank_addr;
   reg [9:0] col_addr;
   reg [2:0] burst_length;
@@ -26,13 +26,16 @@ module sdram(
   reg [1:0] cas_counter;
   reg auto_precharge;
   reg [12:0] mode_register;
+  reg [1:0] precharge_cnt;
+  reg [3:0] bank_en;
 
 
   reg r_en;
   reg w_en;
 
+
   wire [15:0] wstrb;
-  // reg [15:0] wstrb_1;
+  wire is_term;
 
   // Command decoding
   wire cmd_nop = ~cs & ras & cas & we;
@@ -40,18 +43,13 @@ module sdram(
   wire cmd_read = ~cs & ras & ~cas & we;
   wire cmd_write = ~cs & ras & ~cas & ~we;
   wire cmd_precharge = ~cs & ~ras & cas & ~we;
-  wire cmd_refresh = ~cs & ~ras & cas & ~we;
+  wire cmd_refresh = ~cs & ~ras & ~cas & we;
   wire cmd_burst_term = ~cs & ras & cas & ~we;
   wire cmd_load_mode = ~cs & ~ras & ~cas & ~we;
 
+  assign is_term = cmd_burst_term;
   // wstrb
   assign wstrb = { {8{~dqm[1]}}, {8{~dqm[0]}} }; 
-  //always @(posedge clk) begin
-  //  wstrb_1 <= { {8{~dqm[1]}}, {8{~dqm[0]}} }; 
-  //end
-  //always @(posedge clk) begin
-  //  wstrb <= wstrb_1; 
-  //end
 
   // cke
   always @(posedge clk) begin
@@ -59,6 +57,7 @@ module sdram(
       cas_counter <= 2'b0;
       r_en <= 1'b0;
       w_en <= 1'b0;
+      bank_en <= 4'b0;
     end
   end
   // Initialize mode register
@@ -73,14 +72,16 @@ module sdram(
   // Activate command
   always @(posedge clk) begin
     if (cmd_active) begin
-      row_addr <= a;
+      row_addr[ba] <= a;
       bank_addr <= ba;
+      bank_en[ba] <= 1'b1;
     end
   end
 
   // Read command
   always @(posedge clk) begin
     if (cmd_read) begin
+      bank_addr <= ba;
       r_en <= 1'b1;
       col_addr <= a[9:0];
       auto_precharge <= a[10];
@@ -96,22 +97,22 @@ module sdram(
       col_addr <= a[9:0];
       auto_precharge <= a[10];
       burst_counter <= 1;
-      // memory[{bank_addr, row_addr, col_addr}] <= dq;
-      sdram_write({6'b0, bank_addr}, {19'b0, row_addr}, {22'b0, a[9:0]}, wstrb, dq & wstrb);
+      bank_addr <= ba;
+      sdram_write({6'b0, ba}, {19'b0, row_addr[ba]}, {22'b0, a[9:0]}, wstrb, {{16{1'b0}}, dq & wstrb});
     end
   end
 
   // Precharge command
   always @(posedge clk) begin
     if (cmd_precharge) begin
-      // if (a[10]) begin
-      //   // Precharge all banks
-      //   row_addr <= 13'b0;
-      //   bank_addr <= 2'b0;
-      // end else begin
-      //   // Precharge selected bank
-      //   row_addr <= 13'b0;
-      // end
+      if (a[10]) begin
+        // Precharge all banks
+        bank_en <= 4'b0;
+      end else begin
+        // Precharge selected bank
+        bank_en[ba] <= 1'b0;
+      end
+
     end
   end
 
@@ -119,23 +120,27 @@ module sdram(
   always @(posedge clk) begin
     if (cmd_burst_term) begin
       burst_counter <= 0;
+      cas_counter <= 2'b0;
       r_en <= 1'b0;
       w_en <= 1'b0;
-
+      dq_en <= 1'b0;
     end
   end
 
   // Data output
   always @(posedge clk) begin
-    if (r_en && cas_counter == cas_latency - 2'b1 && burst_counter > burst_length) begin
+    if (r_en && cas_counter == cas_latency - 2'b1 && burst_counter > burst_length && !is_term) begin
       r_en <= 1'b0;
       cas_counter <= 2'b0;
       burst_counter <= 3'b0;
       dq_en <= 1'b0;
+      if(auto_precharge) begin
+        bank_en[bank_addr] <= 1'b0;
+        auto_precharge <= 1'b0;
+      end
     end
-    else if (r_en && cas_counter == cas_latency - 2'b1 && burst_counter <= burst_length) begin
-      //  dq_out <= memory[{bank_addr, row_addr, col_addr + burst_counter}];
-      sdram_read({6'b0,bank_addr}, {19'b0, row_addr}, {22'b0, col_addr} + {29'b0, burst_counter}, wstrb, dq_out);
+    else if (r_en && cas_counter == cas_latency - 2'b1 && burst_counter <= burst_length && !is_term) begin
+      sdram_read({6'b0,bank_addr}, {19'b0, row_addr[bank_addr]}, {22'b0, col_addr} + {29'b0, burst_counter}, wstrb, {{16{1'b0}}, dq_out});
       dq_en <= 1;
       burst_counter <= burst_counter + 1;
     end
@@ -150,13 +155,16 @@ module sdram(
 
   // Data input
   always @(posedge clk) begin
-    if (w_en && burst_counter > burst_length) begin
+    if (w_en && burst_counter > burst_length && !is_term && !cmd_write) begin
       w_en <= 1'b0;
       burst_counter <= 3'b0;
+      if(auto_precharge) begin
+        bank_en[bank_addr] <= 1'b0;
+        auto_precharge <= 1'b0;
+      end
     end
-    else if (w_en && burst_counter <= burst_length) begin
-      //  dq_out <= memory[{bank_addr, row_addr, col_addr + burst_counter}];
-      sdram_write({6'b0,bank_addr}, {19'b0, row_addr}, {22'b0, col_addr} + {29'b0, burst_counter}, wstrb, dq & wstrb);
+    else if (w_en && burst_counter <= burst_length && !is_term && !cmd_write && !cmd_burst_term) begin
+      sdram_write({6'b0,bank_addr}, {19'b0, row_addr[bank_addr]}, {22'b0, col_addr} + {29'b0, burst_counter}, wstrb, {{16{1'b0}}, dq & wstrb});
       burst_counter <= burst_counter + 1;
     end
     else if(w_en) begin
@@ -165,14 +173,7 @@ module sdram(
 
     end
   end
-  // NOP
-  always @(posedge clk) begin
-    if(r_en) begin
-
 
-    end
-
-  end
 
   // Data bus control
   assign dq = dq_en ? dq_out : 16'bz;
-- 
2.51.0

