From ef59ec18aad1d2636a889bb100ee42d3f55c4d38 Mon Sep 17 00:00:00 2001
From: Ruomio <1065940593@qq.com>
Date: Thu, 13 Mar 2025 10:58:59 +0800
Subject: [PATCH 11/23] soc: vga achive

---
 perip/vga/vga_top_apb.v | 120 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 120 insertions(+)

diff --git a/perip/vga/vga_top_apb.v b/perip/vga/vga_top_apb.v
index 5c74cfbe..58f0bd78 100644
--- a/perip/vga/vga_top_apb.v
+++ b/perip/vga/vga_top_apb.v
@@ -19,5 +19,125 @@ module vga_top_apb(
   output        vga_vsync,
   output        vga_valid
 );
+// 640 x 480 pixels = 640 x 480 x 3 x 8 = 7372800 bits
+//640x480分辨率下的VGA参数设置
+  parameter    h_frontporch = 96;
+  parameter    h_active = 144;
+  parameter    h_backporch = 784;
+  parameter    h_total = 800;
+
+  parameter    v_frontporch = 2;
+  parameter    v_active = 35;
+  parameter    v_backporch = 515;
+  parameter    v_total = 525;
+ 
+  //像素计数值
+  reg [9:0]    x_cnt;
+  reg [9:0]    y_cnt;
+  reg [9:0]    h_addr;   //提供给上层模块的当前扫描像素点坐标
+  reg [9:0]    v_addr;
+  wire         h_valid;
+  wire         v_valid;
+
+
+  reg [31:0] reg_rdata;
+
+  // 帧缓冲区定义
+  localparam FRAME_BUFFER_SIZE = 640 * 480 * 4 + 8; // 640x480像素，每个像素4字节, 4B: size, 4B: vsync
+  reg [7:0] frame_buffer [0:FRAME_BUFFER_SIZE-1];
+  wire reg_we;
+  wire reg_re;
+
+
+  // APB响应信号
+  assign in_pready = in_psel & in_penable; // 当选中且使能时准备好
+  assign in_pslverr = 1'b0; // 假设没有错误发生
+  assign reg_we  = ~reset & in_psel & ~in_penable &  in_pwrite;
+  assign reg_re  = ~reset & in_psel & ~in_penable & ~in_pwrite;
+
+  assign in_prdata = reg_rdata;
+
+  wire [31:0] relative_addr = in_paddr[31:0] - 32'h21000000;
+  wire [31:0] fb_addr = ((32'd640 * {22'd0,v_addr}) + {22'd0,h_addr})*32'd4 + 32'd8;
+  wire [23:0] vga_data = {frame_buffer[(fb_addr+'d2)], frame_buffer[fb_addr+'d1], frame_buffer[fb_addr]};
+
+  // 根据地址决定读取哪个寄存器或写入帧缓冲区
+  always @(posedge clock) begin
+    if(reset) begin
+      reg_rdata = 32'b0;
+      for(reg [31:0] i=0; i<FRAME_BUFFER_SIZE; i++) begin
+        frame_buffer[i] = 8'b0;
+      end
+      // set size = 640 x 480
+      {frame_buffer[3], frame_buffer[2], frame_buffer[1], frame_buffer[0]} = 32'd640 << 16 | 32'd480;
+    end 
+    else if (reg_we) begin
+      // 写帧缓冲区
+      case(in_pstrb)
+        4'b0001: frame_buffer[relative_addr] = in_pwdata[7:0];
+        4'b0010: frame_buffer[relative_addr+'b1] = in_pwdata[15:8];
+        4'b0100: frame_buffer[relative_addr+'b10] = in_pwdata[23:16];
+        4'b1000: frame_buffer[relative_addr+'b11] = in_pwdata[31:24];
+        4'b0011: begin
+          frame_buffer[relative_addr] = in_pwdata[7:0]; frame_buffer[relative_addr+'b1] = in_pwdata[15:8];
+        end
+        4'b0110: begin
+          frame_buffer[relative_addr+'b1] = in_pwdata[15:8]; frame_buffer[relative_addr+'b10] = in_pwdata[23:16];
+        end
+        4'b1100: begin
+          frame_buffer[relative_addr+'b10] = in_pwdata[23:16]; frame_buffer[relative_addr+'b11] = in_pwdata[31:24];
+        end
+        4'b1111: begin
+          frame_buffer[relative_addr] = in_pwdata[7:0]; frame_buffer[relative_addr+'b1] = in_pwdata[15:8]; 
+          frame_buffer[relative_addr+'b10] = in_pwdata[23:16]; frame_buffer[relative_addr+'b11] = in_pwdata[31:24];
+        end
+        default:
+          $error("error: invalid strobe");
+      endcase
+
+    end else if(reg_re) begin
+      // 读帧缓冲区
+      reg_rdata = {frame_buffer[relative_addr+3], frame_buffer[relative_addr+2], frame_buffer[relative_addr+1], frame_buffer[relative_addr]};
+    end
+    else begin
+      reg_rdata = 32'b0;
+    end
+  end
+
+  always @(posedge reset or posedge clock) //行像素计数
+      if (reset == 1'b1)
+        x_cnt <= 1;
+      else
+      begin
+        if (x_cnt == h_total)
+            x_cnt <= 1;
+        else
+            x_cnt <= x_cnt + 10'd1;
+      end
+
+  always @(posedge clock)  //列像素计数
+      if (reset == 1'b1)
+        y_cnt <= 1;
+      else
+      begin
+        if (y_cnt == v_total & x_cnt == h_total)
+            y_cnt <= 1;
+        else if (x_cnt == h_total)
+            y_cnt <= y_cnt + 10'd1;
+      end
+  //生成同步信号
+  assign vga_hsync = (x_cnt > h_frontporch);
+  assign vga_vsync = (y_cnt > v_frontporch);
+  //生成消隐信号
+  assign h_valid = (x_cnt > h_active) & (x_cnt <= h_backporch);
+  assign v_valid = (y_cnt > v_active) & (y_cnt <= v_backporch);
+  assign vga_valid = h_valid & v_valid;
+  //计算当前有效像素坐标
+  assign h_addr = h_valid ? (x_cnt - 10'd145) : {10{1'b0}};
+  assign v_addr = v_valid ? (y_cnt - 10'd36) : {10{1'b0}};
+  //设置输出的颜色值
+  assign vga_r = vga_data[23:16];
+  assign vga_g = vga_data[15:8];
+  assign vga_b = vga_data[7:0];
 
 endmodule
-- 
2.51.0

