From a7ea30c8e5418a95401cb5aeec79f5ff40decef9 Mon Sep 17 00:00:00 2001
From: Ruomio <1065940593@qq.com>
Date: Fri, 11 Apr 2025 17:44:11 +0800
Subject: [PATCH 82/88] soc: sdram fix read to read bugs

---
 perip/sdram/sdram.v | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/perip/sdram/sdram.v b/perip/sdram/sdram.v
index c847c95b..4d99819b 100644
--- a/perip/sdram/sdram.v
+++ b/perip/sdram/sdram.v
@@ -139,7 +139,7 @@ module sdram(
   always @(posedge clk) begin
     if(r_en && !is_term && id[1] == 1'b0 && row_addr[bank_addr] < 16'h2000) begin
       if(cas_counter == cas_latency - 2'b1) begin
-        if(burst_counter > burst_length) begin
+        if(burst_counter > burst_length && !cmd_read) begin
           r_en <= 1'b0;
           cas_counter <= 2'b0;
           burst_counter <= 3'b0;
@@ -149,6 +149,10 @@ module sdram(
             auto_precharge <= 1'b0;
           end
         end
+        else if(cmd_read) begin
+          sdram_read({6'b0,id}, {6'b0,bank_addr}, {16'b0, row_addr[bank_addr]}, {22'b0, a[9:0]}, wstrb, {{16{1'b0}}, dq_out});
+          dq_en <= 1;
+        end
         else if(burst_counter <= burst_length) begin
           sdram_read({6'b0,id}, {6'b0,bank_addr}, {16'b0, row_addr[bank_addr]}, {22'b0, col_addr} + {29'b0, burst_counter}, wstrb, {{16{1'b0}}, dq_out});
           dq_en <= 1;
@@ -327,7 +331,7 @@ module sdram_word_extend(
   always @(posedge clk) begin
     if(r_en && !is_term && id[1] == 1'b1 && row_addr[bank_addr] >= 16'h2000) begin
       if(cas_counter == cas_latency - 2'b1) begin
-        if(burst_counter > burst_length) begin
+        if(burst_counter > burst_length && !cmd_read) begin
           r_en <= 1'b0;
           cas_counter <= 2'b0;
           burst_counter <= 3'b0;
@@ -337,6 +341,10 @@ module sdram_word_extend(
             auto_precharge <= 1'b0;
           end
         end
+        else if(cmd_read) begin
+          sdram_read({6'b0,id}, {6'b0,bank_addr}, {16'b0, row_addr[bank_addr]}, {22'b0, a[9:0]}, wstrb, {{16{1'b0}}, dq_out});
+          dq_en <= 1;
+        end
         else if(burst_counter <= burst_length) begin
           sdram_read({6'b0,id}, {6'b0,bank_addr}, {16'b0, row_addr[bank_addr]}, {22'b0, col_addr} + {29'b0, burst_counter}, wstrb, {{16{1'b0}}, dq_out});
           dq_en <= 1;
-- 
2.51.0

