From a22c1bc346037f54ffca63971415d1eb76d15559 Mon Sep 17 00:00:00 2001
From: Ruomio <1065940593@qq.com>
Date: Mon, 14 Apr 2025 09:22:35 +0800
Subject: [PATCH 83/88] soc: sdram fix read more than once bug, though not
 effect result

---
 perip/sdram/sdram.v | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/perip/sdram/sdram.v b/perip/sdram/sdram.v
index 4d99819b..c5f73de6 100644
--- a/perip/sdram/sdram.v
+++ b/perip/sdram/sdram.v
@@ -149,10 +149,6 @@ module sdram(
             auto_precharge <= 1'b0;
           end
         end
-        else if(cmd_read) begin
-          sdram_read({6'b0,id}, {6'b0,bank_addr}, {16'b0, row_addr[bank_addr]}, {22'b0, a[9:0]}, wstrb, {{16{1'b0}}, dq_out});
-          dq_en <= 1;
-        end
         else if(burst_counter <= burst_length) begin
           sdram_read({6'b0,id}, {6'b0,bank_addr}, {16'b0, row_addr[bank_addr]}, {22'b0, col_addr} + {29'b0, burst_counter}, wstrb, {{16{1'b0}}, dq_out});
           dq_en <= 1;
-- 
2.51.0

